dali/plugin/plugin_manager.h

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PLUGIN_PLUGIN_MANAGER_H_
+#define DALI_PLUGIN_PLUGIN_MANAGER_H_
+
+#include <string>
+#include <vector>
+#include <mutex>
+#include "dali/common.h"
+
+namespace dali {
+
+class DLL_PUBLIC PluginManager {
+ public:
+    static DLL_PUBLIC PluginManager& Instance();
```

Comment:
If we need a singleton here, ctor, copy ctor, operator=&, move ctor and operator=&& should be deleted. Also, `Instance` could use some of thread-safety mechanisms. 

Anyway, I'm not really fond of singletons, is there any other way, that we could implement this? Maybe other have some better idea? (My best is to have just a global variable, not really nice as well)

Additionally, is there a reason for deleting dtor? I'm asking, cause it's not seen often...

Reply:
Destructor is not deleted. It's private. To make it a complete singleton, constructor should be private, too.

Reply:
About the thread safety, I thought that `static` initialization isn't, but it is, so my bad here